Smooth arcs

kwcantrell opened the pull request:
This PR dynamically adjusts the number of line segments used to draw branch arcs. The number of line segments used to draw an arc is determined by the following formula: `min(60 * (total angle of the arc / pi), 2)` This formula seems to work fairly well on the trees I have tested and also greatly reduced the size of the buffer. On one of the larger trees I tested, the buffer used to draw the tree went from having 39208150 elements to 9230430 elements which is a ~4x smaller!

I also changed the default branch color to a darker grey color after our discussion on Tuesday.

emperor-helper commented:
The following artifacts were built for this PR: [empire-biplot.qzv](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fmchelper.ucsd.edu%2Fdownloads%2Fempress%2F377%2Fempire-biplot.qzv), [empire.qzv](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fmchelper.ucsd.edu%2Fdownloads%2Fempress%2F377%2Fempire.qzv), [empress-tree.qzv](https://view.qiime2.org/visualization/?type=html&src=https%3A%2F%2Fmchelper.ucsd.edu%2Fdownloads%2Fempress%2F377%2Fempress-tree.qzv)

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -398,6 +398,7 @@ define([
 
         this.getLayoutInfo();
         this.centerLayoutAvgPoint();
+        // this.drawTree();
```
```suggestion
        // centerLayoutAvgPoint() calls drawTree(), so no need to call it here
```

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -853,14 +854,19 @@ define([
-                    // of the arc, by arcendangle - arcstartangle radians.
-                    var numSamples = 15;
+                    // An arc will be created for all internal nodes.
+                    // arcs are created by sampling upto 60 small lines along
```
```suggestion
                    // arcs are created by sampling up to 60 small lines along
```

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -853,14 +854,19 @@ define([
+                    // arcs are created by sampling upto 60 small lines along
+                    // the arc spanned by rotating the line (arcx0, arcy0)
+                    // arcendangle - arcstartangle radians. This will create an
+                    // that starts at each internal node's right most child
+                    // and ends on the left most child.
```
```suggestion
                    // that starts at each internal node's rightmost child
                    // and ends on the leftmost child.
```

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -853,14 +854,19 @@ define([
-                    // origin is the root of the tree and endpoint is the start
-                    // of the arc, by arcendangle - arcstartangle radians.
-                    var numSamples = 15;
+                    // An arc will be created for all internal nodes.
+                    // arcs are created by sampling upto 60 small lines along
+                    // the arc spanned by rotating the line (arcx0, arcy0)
+                    // arcendangle - arcstartangle radians. This will create an
+                    // that starts at each internal node's right most child
+                    // and ends on the left most child.
                     var arcDeltaAngle =
                         this.getNodeInfo(node, "arcendangle") -
                         this.getNodeInfo(node, "arcstartangle");
+                    var numSamples = Math.floor(
+                        60 * Math.abs(arcDeltaAngle / Math.PI)
+                    );
+                    numSamples = numSamples > 0 ? numSamples : 2;
                     var sampleAngle = arcDeltaAngle / numSamples;
```
I realize that the code here is already sort of duplicated between this and the thick-line-drawing stuff, but I would recommend abstracting at least some of this logic to a separate function in Empress. This would make changing the arc smoothing in the future easier (e.g. increasing the max samples from 60 to 100, or something).

I imagine writing a function that takes as input `node` and returns `numSamples`, `sampleAngle`, etc. might be one easy way to go about this, but you would probably have a better idea than me about what'd be best.

kwcantrell replied:
I added a function that returns numSamples. I choose not to include sampleAngle because that value is calculated quite differently when drawing the wedges.

fedarko replied:
Sounds good, thanks!

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -853,14 +854,19 @@ define([
+                    var numSamples = Math.floor(
+                        60 * Math.abs(arcDeltaAngle / Math.PI)
+                    );
+                    numSamples = numSamples > 0 ? numSamples : 2;
```
The PR message says that the formula used for the number of line segments is `min(60 * (total angle of the arc / pi), 2)`, but here it looks like it's possible for `numSamples` to be set to 1 if `60 * |angle/pi|` is in the range of `[1, 2)` (e.g. `angle = 0.0795`, in which case `60*(angle/pi) = ~1.5`).

I think this could be fixed by setting `numSamples` after the flooring operation as:

`numSamples = numSamples => 2 ? numSamples : 2;`

kwcantrell replied:
Fixed

fedarko reviewed (changes requested):
A couple suggestions, but nothing huge. This looks great, thanks so much @kwcantrell!

fedarko commented:
Oh, one more thing I forgot -- would it also be worth applying this to wedges for collapsed clades in the circular layout? Not sure how transferable this would be to that, so if you want to defer that to an issue that's fine with me.

kwcantrell commented:
thanks @fedarko! I addressed all of your comments in addition, I fixed an error with the clade collapse algorithm that caused the clades to start in the wrong position (essentially, the algorithm would "skip" the first triangle). The pr should be good to go.

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -742,6 +742,22 @@ define([
     };
 
+    /**
+     * Returns the number of lines/trianges to approximate an arc/wedge given
```
```suggestion
     * Returns the number of lines/triangles to approximate an arc/wedge given
```

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -742,6 +742,22 @@ define([
+        var numSamples = Math.floor(
+            60 * Math.abs(totalAngle / Math.PI)
+        );
+        numSamples = numSamples >= 2 ? numSamples : 2;
+        return numSamples;
```
```suggestion
        return numSamples >= 2 ? numSamples : 2;
```
(this is subjective, feel free to ignore it)

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -853,14 +868,16 @@ define([
+                    // An arc will be created for all internal nodes.
+                    // arcs are created by sampling up to 60 small lines along
+                    // the arc spanned by rotating the line (arcx0, arcy0)
+                    // arcendangle - arcstartangle radians. This will create an
+                    // that starts at each internal node's rightmost child
+                    // and ends on the leftmost child.
```
```suggestion
                    // arcendangle - arcstartangle radians. This will create an
                    // arc that starts at each internal node's rightmost child
                    // and ends on the leftmost child.
```

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -1245,14 +1262,16 @@ define([
-                    // arc, by arcendangle - arcstartangle radians.
-                    var numSamples = 15;
+                    // An arc will be created for all internal nodes.
+                    // arcs are created by sampling upto 60 small lines along
+                    // the arc spanned by rotating the line (arcx0, arcy0)
+                    // arcendangle - arcstartangle radians. This will create an
+                    // that starts at each internal node's right most child
+                    // and ends on the left most child.
```
```suggestion
                    // See getCoords() for details on how arcs are drawn.
```
(To cut down on redundant documentation)

fedarko commented on empress/support_files/js/empress.js:
```diff
@@ -2630,11 +2648,12 @@ define([
+            // create triangles to approximate sector
+            var numSamples = this._numSampToApproximate(totalAngle);
+            var deltaAngle = totalAngle / numSamples;
+            cos = Math.cos(0);
+            sin = Math.sin(0);
```
I doubt that these would be a bottleneck, but we may as well avoid extra work where possible. If there are a ton of collapsed clades this _might_ save some time.
```suggestion
            cos = 1; // Math.cos(0)
            sin = 0; // Math.sin(0)
```

fedarko reviewed (changes requested):
Thanks @kwcantrell! Awesome catch with the clade collapsing -- I can verify that the collapsing angles look more consistent now.

Old:

![old](https://user-images.githubusercontent.com/4177727/93388524-b4402a00-f81f-11ea-9d34-44f3d2367b6a.png)

New:

![new](https://user-images.githubusercontent.com/4177727/93388515-b1ddd000-f81f-11ea-9a5c-19485d811d0a.png)

I have a few small suggestions on the latest version of the code, but they should all be pretty tiny. Looking forwards to getting this merged in! :curling_stone:

kwcantrell commented:
thanks @fedarko!  I addressed your remaining comments so it should be good to merge.

fedarko commented:
Thanks @kwcantrell!
